File: 08/prog.py
import sys

screen = [[False for x in range(6)] for y in range(50)]
count = 0

while True:
    line = sys.stdin.readline().rstrip()
    if not line: break

    parts = line.split()
    if parts[0] == "rect":
        (x, y) = parts[1].split("x")
        x = int(x)
        y = int(y)

        for i in range(x):
            for j in range(y):
                screen[i][j] = True

    elif parts[0] == "rotate":
        assert parts[3] == "by"
        n = int(parts[4])

        if parts[1] == "row":
            (_, y) = parts[2].split("=")
            y = int(y)

            row = [False] * 50
            for i in range(50):
                dest = i + n
                if dest >= 50:
                    dest -= 50
                row[dest] = screen[i][y]
            for i in range(50):
                screen[i][y] = row[i]

        elif parts[1] == "column":
            (_, x) = parts[2].split("=")
            x = int(x)

            col = [False] * 6
            for i in range(6):
                dest = i + n
                if dest >= 6:
                    dest -= 6
                col[dest] = screen[x][i]
            for i in range(6):
                screen[x] = col

        else:
            raise Exception("invalid rotation {}".format(parts[1]))

    else:
        raise Exception("invalid command {}", parts[0])

for y in range(6):
    for x in range(50):
        if screen[x][y]:
            print("#", end="")
            count += 1
        else:
            print(".", end="")
    print()
print(count)

File: 08/prog.rs
fn print_screen(screen: &[[bool;6];50]) {
    for y in 0..6 {
        for x in 0..50 {
            if screen[x][y] {
                print!("#");
            } else {
                print!(".");
            }
        }
        println!("");
    }
}

fn main() {
    let animate = std::env::args()
        .skip(1)
        .next()
        .and_then(|x| Some(x == "--animate"))
        .unwrap_or(false);

    let mut count = 0;
    let mut screen = [[false;6];50];

    let mut line = String::new();
    while let Ok(n) = std::io::stdin().read_line(&mut line) {
        if n == 0 { break; }

        {
            let parts = line.trim().split(" ").collect::<Vec<&str>>();

            match parts[0] {
                "rect" => {
                    let mut coords = parts[1].split("x");
                    let x: usize = coords.next().unwrap().parse().unwrap();
                    let y: usize = coords.next().unwrap().parse().unwrap();

                    for i in 0..x {
                        for j in 0..y {
                            screen[i][j] = true;
                        }
                    }
                },
                "rotate" => {
                    assert_eq!("by", parts[3]);
                    let n: usize = parts[4].parse().unwrap();

                    match parts[1] {
                        "row" => {
                            let y: usize = parts[2].split("=").skip(1).next().unwrap().parse().unwrap();

                            let mut row = [false;50];
                            for i in 0..50 {
                                let mut dest = i+n;
                                if dest >= 50 {
                                    dest -= 50;
                                }
                                row[dest] = screen[i][y];
                            }
                            for i in 0..50 {
                                screen[i][y] = row[i];
                            }
                        },
                        "column" => {
                            let x: usize = parts[2].split("=").skip(1).next().unwrap().parse().unwrap();

                            let mut column = [false;6];
                            for i in 0..6 {
                                let mut dest = i+n;
                                if dest >= 6 {
                                    dest -= 6;
                                }
                                column[dest] = screen[x][i];
                            }
                            screen[x] = column;
                        },
                        _ => panic!("unknown rotate: {}", parts[1])
                    }
                }
                _ => panic!("unknown instruction: {}", parts[0])
            }
        }

        if animate {
            print!("\x1b[2J\x1b[H"); // clear screen and set cursor to 0,0
            print_screen(&screen);
            std::thread::sleep(std::time::Duration::from_millis(50));
        }
        line.clear();
    }
    if !animate {
        print_screen(&screen);
    }
    for y in 0..6 {
        for x in 0..50 {
            if screen[x][y] {
                count += 1;
            }
        }
    }
    println!("{}", count);
}
